Automattic/pocket-casts-ios
# Player: custom transition

Address the remaining of https://github.com/Automattic/pocket-casts-ios/discussions/1037

This PR builds upon the last one and adds a custom transition from miniplayer to the full player. You can see the demo below in the normal speed and low speed.

https://github.com/Automattic/pocket-casts-ios/assets/7040243/dffbb573-9442-40f2-9ce8-cfcc0b3a99a2

I tried to tweak the animations to feel as good and smooth as possible. The player dismissal logic was also changed, in a way that if the user swipes down aggressively, the player will be dismissed fast.

## To test

I highly recommend testing this on real devices as it will help you get the real feeling of the new transition.

1. Run the app
2. Play any podcast
3. Tap the miniplayer
4. ✅ The full player should appear
5. Swipe down
6. ✅ The full player should dismiss
7. Reopen the player
8. Drag down to 15 or 20% of the screen and let it go
9. ✅ Check that the player dismiss and the artwork moves to the artwork in the miniplayer
10. Reopen the full player
11. Swipe down as quickly as you can
12. ✅ The player should dismiss as quickly as you swiped
13. Play a video podcast
14. Dismiss the player
15. ✅ There are no artwork animations and the full player is dismissed
16. Tap the miniplayer
17. ✅ The full player should open without the artwork animation
18. Open the full player and swipe tabs then dismiss it
19. ✅ The animation should perform

Please, test on small and big devices. :)

## Breakdown

It seems a simple animation but there's a lot going on. :) In terms of animation what is happening is:

1. The artwork moves from the small artwork to the big artwork, if the current episode **is not** a videocast
2. The miniplayer "grows" to the size of the full player
3. The full player appears with a fade in animation
4. When presenting the full player, we're using easy in-out. This means that the animation starts slowly, gets faster, and then slows down again.
5. However, when dismissing we use easy out, the animation starts fast and then slows down. I played a lot with that and this was the most natural result.

## Possible improvements

There are a few improvements/ideas we can also make so this is even better:

1. We can also hide the tab bar as the miniplayer is "growing"
2. We could do a bounce animation when dismissing
3. Sometimes it seems that the swipe left/right gestures get in the way of the swipe down gesture. This isn't a behavior added in this PR but it's something we can improve so the gesture is even more natural

## Checklist
